### ISSUE TITLE
arch: cognitive observability layer — internal state exposure, transformation buffer, and presentation

### ISSUE THREAD

**Issue by Peleke** [Peleke/buildlog-template#135]
## Context

This is NOT infrastructure observability. This is NOT LLM-call tracing. This is **cognitive observability** — exposing the internal reasoning and belief state of an agent system as structured, observable signals.

Existing tools observe the wrong things for our purposes:
- **LangSmith**: traces LLM calls, tool invocations, token costs
- **Logfire**: OTLP spans with AI-specific attributes (gen_ai.* conventions)
- **Mastra**: agent execution traces with noise filtering
- **Dash0**: AI agents that consume infrastructure telemetry
- **DeepEval**: post-hoc LLM output evaluation (faithfulness, hallucination)
- **Guardrails AI**: runtime output enforcement (safety, formatting)

All of these observe **what the system did**. None of them observe **what the system believes, how confident it is, or how its cognition is changing over time**.

Our observability layer observes cognition. It has three parts:
1. **Internal state** — raw, high-dimensional, honest
2. **Buffer** — transformation space where lossy compression occurs
3. **Presentation** — low-dimensional output (dashboards for humans, signal stream for interoception)

The buffer is the novel piece. Internal state → presentation is lossy, like how internal emotional state → facial expressions is lossy. The buffer defines HOW that compression happens, and both interfaces (internal state schema, presentation schema) are first-class.

### Related Architecture
- **Mnemosyne** (learning layer): primary source of cognitive state (posteriors, credit, convergence)
- **Short-term memory** (#133): memory fullness, entity staleness, recall patterns
- **Cadence**: heartbeat provides the clock; observability emits on each beat
- **Interoception** (affect): CONSUMES observability output (inward face) to anticipate state
- **OpenClaw green module**: environmental tracking — built into this layer as a non-cognitive but co-emitted signal

### Build Incrementally
The layer starts minimal (Thompson Sampling dashboards) and doesn't block later sophistication (Pyro predictive modeling, RL-trained presentation policies). Each increment adds observable cognitive dimensions without breaking the interface.

---

### From Logfire (Transport + Data Model)
- OTLP as transport standard — our cognitive signals ride on standard spans
- `logfire-api` zero-dependency shim pattern — instrument without forcing installation
- Pending spans — see cognitive state DURING reasoning, not just after
- In-span metric aggregation — parent spans inherit descendant metrics automatically
- JSON schema for type preservation on complex attributes

### From LangSmith (Trace Hierarchy)
- `@traceable` decorator + context variable nesting — automatic parent-child hierarchy
- `RunTree` for manual hierarchy construction
- Feedback/evaluation as first-class objects attached to traces
- `dotted_order` for efficient hierarchical querying

### From Mastra (Noise Filtering)
- Non-exportable internal spans — framework plumbing filtered out by default
- `configSelector` for runtime trace routing (dev vs prod vs interoception)
- Multi-destination export from single trace (dashboard + interoception + archive)
- SensitiveDataFilter as built-in processor

### From Dash0 (Agent Personas for Observability)
- Specialized agent personas consuming telemetry (future: interoception as a consumer persona)
- Integrated Context — 13+ context types auto-inferred from current state
- Open-standard outputs (everything portable, no lock-in)

### From DeepEval (Evaluation Metrics)
- `BaseMetric` interface: `measure(test_case) → score + reason`
- G-Eval for natural language criteria (could define cognitive health criteria in prose)
- DAG metrics for deterministic decision-tree evaluation
- `@observe` decorator for production tracing with attached metrics

### From Guardrails AI (Enforcement)
- `Guard` + `Validator` pattern — could enforce cognitive constraints (e.g., "confidence must not collapse below threshold")
- `OnFailAction` remediation patterns (reask, fix, alert)
- Validator composition (`.use_many()`)

### From OpenClaw Green Module (Co-Emitted Signals)
- SQLite storage with WAL journaling
- Lazy-init API handler pattern (`createGreenApiHandler`)
- Chart.js dashboard generated on-the-fly (no filesystem writes)
- Regulatory export formats (GHG Protocol, TCFD, CDP, ISO 14064)
- Confidence-based uncertainty bounds
- Identical module structure to learning API — both follow: types → store → calculator → capture → api → dashboard-html

---

### Part 1: Internal State (Raw Cognitive Signals)

These are the things we want to observe. Each is a **signal source** with a defined schema:

#### From Mnemosyne (Learning Layer)
| Signal | Schema | Update Frequency |
|--------|--------|-----------------|
| Arm posteriors | `{arm_id, context, alpha, beta, expected_value, ci_lower, ci_upper, pulls}` | Per gauntlet issue / per run |
| Credit attribution | `{arm_id, issue_id, category, reward}` | Per gauntlet issue |
| Citation validation | `{total, valid, hallucinated, hallucination_rate}` | Per gauntlet run |
| Convergence | `{arm_id, posterior_entropy, entropy_delta, converged: bool}` | Per update |
| Exploration rate | `{context, explore_fraction, exploit_fraction}` | Per decision |

#### From Short-Term Memory
| Signal | Schema | Update Frequency |
|--------|--------|-----------------|
| Capacity | `{session_id, items, max_items, fullness_ratio, tokens, token_budget}` | Per ingest |
| Entity state | `{entity_id, fact_count, last_updated, staleness_seconds}` | Per track |
| Recall performance | `{strategy, query, latency_ms, items_returned, hit_rate}` | Per recall |
| Promotion events | `{item_id, trigger, target, promoted_at}` | Per promote |
| Compaction events | `{strategy, items_kept, items_dropped, items_promoted, tokens_saved}` | Per compact |

#### From Cadence (Clock)
| Signal | Schema | Update Frequency |
|--------|--------|-----------------|
| Heartbeat | `{beat_number, timestamp, block, signals_emitted}` | Per heartbeat |
| Block transition | `{from_block, to_block, duration_seconds}` | Per transition |

#### From Green (Environmental)
| Signal | Schema | Update Frequency |
|--------|--------|-----------------|
| Carbon trace | `{run_id, total_co2_grams, water_ml, tokens, confidence}` | Per run |
| Cumulative impact | `{total_co2, equivalents, intensity_per_million_tokens}` | On query |

#### Meta-Cognitive (About Observability Itself)
| Signal | Schema | Update Frequency |
|--------|--------|-----------------|
| Signal volume | `{signals_per_heartbeat, by_source, buffer_pressure}` | Per heartbeat |
| Transformation loss | `{source_dimensions, presentation_dimensions, information_loss_estimate}` | Per render |

### Part 2: Buffer (Transformation Space)

The buffer sits between internal state and presentation. It defines HOW cognitive state is compressed for output. Transformations are pluggable strategies (like recall strategies in STM):

#### Buffer Interface
```
transform(internal_state) → presentation_state
    Lossy compression of raw cognitive signals into presentable form.
    
register_transform(source, strategy)
    Attach a transformation strategy to a signal source.
    
get_buffer_state() → BufferState
    Current buffer contents (pre-transformation).
    Useful for interoception to read raw state.
```

#### Transformation Strategies

**For arm posteriors → Power Rankings:**
- Input: full Beta(alpha, beta) per arm per context
- Output: ranked list with scores + confidence bars
- Loss: individual posterior shapes compressed to point estimates + intervals

**For citation stats → Health Indicator:**
- Input: per-run citation validation details
- Output: single hallucination rate trend (green/yellow/red)
- Loss: individual hallucinated IDs, per-issue detail

**For memory capacity → Pressure Gauge:**
- Input: items, tokens, staleness per entity, recall miss rates
- Output: single pressure score (0-1) with trend arrow
- Loss: per-entity detail, per-recall latencies

**For convergence → Learning Curve:**
- Input: per-arm entropy time series
- Output: aggregate convergence plot + "time to convergence" estimate
- Loss: per-arm detail collapsed to population statistics

#### Key Design Property
The buffer is where interoception diverges from dashboards:
- **Dashboard adapter**: high loss, pretty, human-readable
- **Interoception adapter**: low loss, structured, machine-readable
- Both read from the same buffer; different transformation strategies

This means interoception gets a richer view of internal state than humans do, which is architecturally correct — you don't show everything on your face that you feel internally.

#### Outward Face: Human Dashboards

**MVP Dashboard (Phase 1):**
- Arm Power Rankings (sorted by expected value, with Beta distribution sparklines)
- Hallucination Rate trend (line chart, per-gauntlet-run)
- Persona ROI (aggregate arm scores by persona)
- Rule × Category heatmap (arms on Y, categories on X, color = score)
- Green impact summary (co-emitted, always visible)

**Technology:** Streamlit or Panel for Phase 1 (Python, reads SQLite directly). OpenClaw-style on-the-fly HTML for embedded dashboards later.

**Phase 2 Dashboard Additions:**
- Credit Attribution Sankey (persona → rule → category → severity)
- Convergence curves (entropy over time per arm)
- Memory pressure gauge + promotion history
- Recall strategy comparison (which strategies produce better outcomes?)

#### Inward Face: Interoception Feed

The interoception layer reads the buffer and derives:
- **Anticipation signals**: "posterior entropy is decreasing → exploration will decrease next heartbeat"
- **Anomaly detection**: "hallucination rate spiked → something changed in the prompt or rules"
- **Capacity pressure**: "memory is 85% full → promote or compact within 2 heartbeats"
- **Homeostatic targets**: "exploration rate should be between 10-30%; currently at 5% → trigger rebalance"

**Future (Phase 3+):** Interoception's reading of observability becomes the state representation for an RL agent that learns WHEN to act on cognitive signals. DoWhy provides counterfactual reasoning: "what would have happened if we'd promoted that memory earlier?" Pyro provides the probabilistic programming substrate for predictive modeling of cognitive state evolution.

---

## Implementation: MCP Server

Same form factor as Mnemosyne and STM — an MCP server that owns its state and speaks a tool interface.

**MCP Tools:**
```
cog_observe(source, signal, metadata?)
    Record a cognitive signal from any source.
    Sources: mnemosyne, stm, cadence, green, meta

cog_buffer(filter?)
    Read current buffer state (pre-transformation).
    Used by interoception for low-loss access.

cog_dashboard(view?, format?)
    Render a dashboard view.
    Views: power_rankings, hallucination_trend, persona_roi, heatmap, green, convergence
    Formats: html, json, csv

cog_transform(source, strategy?)
    Apply a transformation strategy to a signal source.
    Returns presentation-ready output.

cog_status()
    Overall cognitive health summary.
    Returns: learning_rate, memory_pressure, hallucination_trend, 
             exploration_rate, convergence_estimate, green_impact
```

**Storage:** SQLite (same pattern as OpenClaw green + learning modules)
- `cognitive_signals` table (append-only, partitioned by source)
- `buffer_state` table (current state, updated per heartbeat)
- `presentation_cache` table (rendered dashboards, TTL-based)

---

### Phase 1: Thompson Sampling Dashboard (Minimal)
- Signal sources: Mnemosyne posteriors, citation stats, green traces
- Buffer: identity transform (no compression, just relay)
- Presentation: Streamlit dashboard with power rankings + hallucination trend + green summary
- No interoception yet

### Phase 2: Multi-Source + Buffer Strategies
- Add STM signals (capacity, recall performance, promotions)
- Add Cadence signals (heartbeat, block transitions)
- Implement buffer with pluggable transformation strategies
- Add heatmap, Sankey, convergence views
- Interoception can read buffer (low-loss) but doesn't act yet

### Phase 3: Interoception Integration
- Define homeostatic targets (exploration rate range, memory pressure thresholds)
- Interoception reads buffer → derives anticipation signals
- Anomaly detection on cognitive signal streams
- Interoception signals feed back to Mnemosyne (adjust exploration) and STM (trigger promotion)

### Phase 4: Predictive Modeling
- Pyro probabilistic programs for cognitive state evolution
- DoWhy counterfactual reasoning ("what if we'd explored more?")
- RL-trained policy for when/what to surface from buffer to presentation
- The system learns how to express its own state

---

## Open Questions

1. **Signal granularity**: Per-heartbeat vs per-event? Per-heartbeat is simpler and aligns with Cadence; per-event is richer but noisier.
2. **Buffer capacity**: Should the buffer accumulate history or only hold current state? History enables trend detection but costs storage.
3. **Transformation as learning target**: Should the buffer's transformation strategies themselves be arms in Mnemosyne? (The system learns how to express itself.)
4. **Green as co-emission vs separate**: Should green signals always co-emit with cognitive signals, or remain queryable separately?
5. **Naming**: Candidates for the observability layer: `mirror`, `cortex-display`, `phenotype`, `visage`, or just `obs`.

---

## References

- LangSmith SDK: `@traceable`, RunTree hierarchy, evaluation framework
- Pydantic Logfire: OTLP wrapper, pending spans, in-span metric aggregation, `logfire-api` shim
- Mastra: noise filtering (non-exportable spans), `configSelector`, multi-destination export
- Dash0 Agent0: specialized agent personas consuming telemetry, Integrated Context
- DeepEval: `BaseMetric` interface, G-Eval natural language criteria, `@observe` decorator
- Guardrails AI: `Guard` + `Validator`, `OnFailAction` remediation
- OpenClaw green: SQLite + WAL, lazy API handler, on-the-fly Chart.js dashboard, regulatory exports
- OpenClaw learning: Thompson Sampling dashboard, arm posteriors API, identical module structure to green
- Singularity SAD v2: observability as "facial expressions" metaphor
- Short-term memory: #133
